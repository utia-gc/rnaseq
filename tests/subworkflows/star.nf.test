nextflow_workflow {

    name "Test Workflow Star"
    script "subworkflows/star.nf"
    workflow "Star"

    def refs = new RefsEnsembl()

    test("Star builds index and aligns reads -- SE reads.") {
        // instantiate helper objects
        def readsLane1 = new ReadsSELane1()

        when {
            params {
                publishDirReports = "${outputDir}/reports"
                publishDirData    = "${outputDir}/data"
            }
            workflow {
                """
                input[0] = [
                    [
                        sampleName:   "${readsLane1.getSampleName()}",
                        sampleNumber: "${readsLane1.getSampleNumber()}",
                        lane:         "${readsLane1.getLane()}",
                        readType:     "${readsLane1.getReadType()}",
                        rgFields:     [
                            ID: "${readsLane1.getRGFields().get('ID')}",
                            SM: "${readsLane1.getRGFields().get('SM')}",
                            LB: "${readsLane1.getRGFields().get('LB')}",
                            PL: "${readsLane1.getRGFields().get('PL')}",
                        ]
                    ],
                    file("${projectDir}/${readsLane1.getR1()}"),
                    file("${projectDir}/${readsLane1.getR2()}")
                ]
                input[1] = file("${projectDir}/${refs.getGenome()}")
                input[2] = file("${projectDir}/${refs.getGTF()}")
                """
            }
        }

        then {
            // test workflow status
            assert workflow.success

            // test BAM
            with(workflow.out.alignments.get(0)) {
                assert size() == 2
                // test BAM file
                assert get(1) ==~ /^.*\/${readsLane1.getStemName()}_Aligned\.out\.sam/
                // test SAM contains read group line
                assert FileHelpers.grepq(path(get(1)), '@RG')
                assert FileHelpers.grepq(path(get(1)), readsLane1.getRGLine())
            }

            // test reproducibility
            assert snapshot(workflow.out).match()
        }

    }

    test("Star builds index and aligns reads -- PE reads.") {
        // instantiate helper objects
        def readsLane1 = new ReadsPELane1()

        when {
            params {
                publishDirReports = "${outputDir}/reports"
                publishDirData    = "${outputDir}/data"
            }
            workflow {
                """
                input[0] = [
                    [
                        sampleName:   "${readsLane1.getSampleName()}",
                        sampleNumber: "${readsLane1.getSampleNumber()}",
                        lane:         "${readsLane1.getLane()}",
                        readType:     "${readsLane1.getReadType()}",
                        rgFields:     [
                            ID: "${readsLane1.getRGFields().get('ID')}",
                            SM: "${readsLane1.getRGFields().get('SM')}",
                            LB: "${readsLane1.getRGFields().get('LB')}",
                            PL: "${readsLane1.getRGFields().get('PL')}",
                        ]
                    ],
                    file("${projectDir}/${readsLane1.getR1()}"),
                    file("${projectDir}/${readsLane1.getR2()}")
                ]
                input[1] = file("${projectDir}/${refs.getGenome()}")
                input[2] = file("${projectDir}/${refs.getGTF()}")
                """
            }
        }

        then {
            // test workflow status
            assert workflow.success

            // test BAM
            with(workflow.out.alignments.get(0)) {
                assert size() == 2
                // test BAM file
                assert get(1) ==~ /^.*\/${readsLane1.getStemName()}_Aligned\.out\.sam/
                // test SAM contains read group line
                assert FileHelpers.grepq(path(get(1)), '@RG')
                assert FileHelpers.grepq(path(get(1)), readsLane1.getRGLine())
            }

            // test reproducibility
            assert snapshot(workflow.out).match()
        }

    }

}
