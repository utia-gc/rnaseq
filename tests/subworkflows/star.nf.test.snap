{
    "Star builds index and aligns reads -- PE reads.": {
        "content": [
            {
                "0": [
                    [
                        {
                            "sampleName": "SRR6924569",
                            "sampleNumber": "1",
                            "lane": "001",
                            "readType": "paired",
                            "rgFields": {
                                "ID": "SRR6924569.001",
                                "SM": "SRR6924569",
                                "LB": "SRR6924569",
                                "PL": "ILLUMINA"
                            }
                        },
                        "SRR6924569_L001_Aligned.out.sam:md5,645f250c180cc36c5898053e2e4eb6b5"
                    ]
                ],
                "alignments": [
                    [
                        {
                            "sampleName": "SRR6924569",
                            "sampleNumber": "1",
                            "lane": "001",
                            "readType": "paired",
                            "rgFields": {
                                "ID": "SRR6924569.001",
                                "SM": "SRR6924569",
                                "LB": "SRR6924569",
                                "PL": "ILLUMINA"
                            }
                        },
                        "SRR6924569_L001_Aligned.out.sam:md5,645f250c180cc36c5898053e2e4eb6b5"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-10-14T14:12:10.149284233"
    },
    "Star builds index and aligns reads -- SE reads.": {
        "content": [
            {
                "0": [
                    [
                        {
                            "sampleName": "SRR1066657",
                            "sampleNumber": "3",
                            "lane": "001",
                            "readType": "single",
                            "rgFields": {
                                "ID": "SRR1066657.001",
                                "SM": "SRR1066657",
                                "LB": "SRR1066657",
                                "PL": "ILLUMINA"
                            }
                        },
                        "SRR1066657_L001_Aligned.out.sam:md5,7399b4b60d585f02ae3db8a3e3867ea9"
                    ]
                ],
                "alignments": [
                    [
                        {
                            "sampleName": "SRR1066657",
                            "sampleNumber": "3",
                            "lane": "001",
                            "readType": "single",
                            "rgFields": {
                                "ID": "SRR1066657.001",
                                "SM": "SRR1066657",
                                "LB": "SRR1066657",
                                "PL": "ILLUMINA"
                            }
                        },
                        "SRR1066657_L001_Aligned.out.sam:md5,7399b4b60d585f02ae3db8a3e3867ea9"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-10-14T14:08:39.030618946"
    }
}